geopandas/tools/overlay.py

```diff
@@ -176,3 +177,177 @@ def overlay(df1, df2, how, use_sindex=True):
 
     # Return geodataframe with new indices
     return GeoDataFrame(collection, index=range(len(collection)))
+
+
+def overlay_intersection(df1, df2):
+    """
+    Overlay Intersection operation used in overlay function
+    """
+    # Spatial Index to create intersections
+    spatial_index = df2.sindex
+    bbox = df1.geometry.apply(lambda x: x.bounds)
+    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
+    # Create pairs of geometries in both dataframes to be intersected
+    nei = []
+    for i, j in sidx.items():
+        for k in j:
+            nei.append([i, k])
+    if nei != []:
+        pairs = pd.DataFrame(nei, columns=['idx1','idx2'])
+        left = df1.geometry.take(pairs['idx1'].values)
+        left.reset_index(drop=True, inplace=True)
+        right = df2.geometry.take(pairs['idx2'].values)
+        right.reset_index(drop=True, inplace=True)
+        intersections = left.intersection(right).buffer(0)
+
+        # only keep actual intersecting geometries
+        pairs_intersect = pairs[~intersections.is_empty]
+        geom_intersect = intersections[~intersections.is_empty]
+
+        # merge data for intersecting geometries
+        dfinter = pairs_intersect.merge(
+            df1.drop(df1._geometry_column_name, axis=1),
+            left_on='idx1', right_index=True)
+        dfinter = dfinter.merge(
+            df2.drop(df2._geometry_column_name, axis=1),
+            left_on='idx2', right_index=True, suffixes=['_1', '_2'])
+
+        return GeoDataFrame(dfinter, geometry=geom_intersect, crs=df1.crs)
+    else:
+        return GeoDataFrame([], columns=list(set(df1.columns).union(df2.columns)), crs=df1.crs)
+
+
+def overlay_difference(df1, df2):
+    """
+    Overlay Difference operation used in overlay function
+    """
+    # Spatial Index to create intersections
+    spatial_index = df2.sindex
+    bbox = df1.geometry.apply(lambda x: x.bounds)
+    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
+    # Create differences
+    new_g = []
+    for geom, neighbours in zip(df1.geometry, sidx):
+        new = reduce(lambda x, y: x.difference(y).buffer(0),
+                     [geom] + list(df2.geometry.iloc[neighbours]))
+        new_g.append(new)
```

Comment:
@ozak similar here: I kept the reduce method, but changed it to a explicit for loop instead of the `apply` call. This makes it a little bit more efficient (in the apply, each row is converted into an object Series, which gives some overhead, although minor compared to the actual `difference` calls), and also a bit more readable I think.

This loop part would actually be nice (and rather easy) to move to cython in the future for further speed-up